net/ip.c

```diff
@@ -0,0 +1,491 @@
+ */
+static struct m0_mutex m0_net_ip_mutex = {};
+
+/**
+ * This function convert the ip to hostname/FQDN format.
+ * Here ip is string containing ip address with numbes-and-dot notation.
+ */
+static int m0_net_ip_to_hostname(const char *ip, char *hostname)
+{
+	struct hostent *he;
+	struct in_addr  addr;
+	int             rc;
+
+	inet_aton(ip, &addr);
+	he = gethostbyaddr(&addr, sizeof(addr), AF_INET);
+	if (he == NULL) {
+		M0_LOG(M0_ERROR, "gethostbyaddr err=%d for %s",
+		       h_errno, (char*)ip);
+		/* Return error code for gethostbyaddr failure */
+		return M0_ERR(-h_errno);
+	}
+	M0_LOG(M0_DEBUG, "ip: %s to hostname: %s\n",ip, he->h_name);
+	rc = snprintf(hostname, M0_NET_IP_STRLEN_MAX, "%s", he->h_name);
+
+	if (rc <= 0 || rc >= M0_NET_IP_STRLEN_MAX)
+		return M0_ERR(-EINVAL);
+
+	return M0_RC(0);
+}
+
+
+static int parse_prefix(const char *ep_name, const char **prefixes,
+			 int nr_prefixes, int *index, int *shift)
+{
+	int i;
+
+	for (i = 0; i < nr_prefixes; ++i) {
+		if (prefixes[i] != NULL) {
+			*shift = strlen(prefixes[i]);
+			if (strncmp(ep_name, prefixes[i], *shift) == 0) {
+				*index = i;
+				break;
+			}
+		}
+	}
+
+	if (i >= nr_prefixes)
+		return M0_ERR(-EINVAL);
+
+	return 0;
+
+}
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	long int     portnum;
+	int          shift = 0;
+	int          family = 0;
+	int          type = 0;
+	int          rc;
+	char         ip[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[M0_NET_IP_PORTLEN_MAX] = {};
+	char        *at;
+	const char  *ep_name = name;
+	char        *end;
+
+	rc = parse_prefix(ep_name, ip_family, ARRAY_SIZE(ip_family), &family,
+			  &shift);
+	if (rc != 0 || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+
+	rc = parse_prefix(ep_name, ip_protocol, ARRAY_SIZE(ip_protocol), &type,
+		     &shift);
+	if (rc != 0 || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+
+	at = strchr(ep_name, '@');
+	if (at == NULL)
+		return M0_ERR(-EINVAL);
+	else {
+		at++;
+		if (at == NULL || !isdigit(at[0]))
+			return M0_ERR(-EINVAL);
+		strcpy(port, at);
+		portnum = strtol(port, &end, 10);
+		if (portnum > M0_NET_IP_PORT_MAX)
+			return M0_ERR(-EINVAL);
+		addr->nia_n.nip_port = (uint16_t)portnum;
+	}
+
+	rc = m0_net_hostname_to_ip((char *)ep_name, ip,
+				   &addr->nia_n.nip_format);
+	if (rc == 0)
+		inet_pton(family == M0_NET_IP_AF_INET ? AF_INET : AF_INET6,
+			  ip, &addr->nia_n.nip_ip_n.sn[0]);
+
+	M0_ASSERT(strlen(name) < ARRAY_SIZE(addr->nia_p));
+	strcpy(addr->nia_p, name);
```

Comment:
We are restricting length to M0_NET_IP_STRLEN_MAX, that can cause problem. The user can allocate space for address and we can use pointer here? This approach will help to optimize string comparison for libfab_buf_add for the case of M0_NET_QT_MSG_SEND e.g. we can do pointer comparison if success then match else do full string comparison.

Reply:
Added optimization for this in 839a562 commit.